### ISSUE TITLE
[Bug]: Milvus 2.4.x GPU version crashes on machines without a GPU

### ISSUE THREAD

**Issue by ashkrisk** [milvus-io/milvus#36894]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.4.12
- Deployment mode(standalone or cluster): both
- MQ type(rocksmq, pulsar or kafka): rocksmq in standalone and kafka in cluster
- SDK version(e.g. pymilvus v2.0.0rc2): NA
- OS(Ubuntu or CentOS): CentOS
- CPU/Memory: 
- GPU: Not attached
- Others:
```

### Current Behavior

When Milvus is built with support for GPU indexes, the resulting binary crashes when attempting to run it on machines without a GPU.

When deployed in standalone mode, the entire process crashes. When deployed in cluster mode, the querynode and indexnodes crash.

### Expected Behavior

The GPU-enabled Milvus binary should run without crashing even on machines without a GPU. Of course, GPU indexes cannot be used, but CPU indexes should still be usable.

### Steps To Reproduce

```markdown
1. Build Milvus with GPU support.
2. Deploy it with `milvus run standalone` (or simply `querynode` or `indexnode`).
3. Observe that Milvus crashes with an error mentioned above.
```

### Milvus Log

W20241008 10:02:22.335795    86 ExceptionTracer.cpp:187] Invalid trace stack for exception of type: raft::cuda_error
terminate called after throwing an instance of 'raft::cuda_error'
  what():  CUDA error encountered at: file=/root/milvus/cmake_build/thirdparty/knowhere/knowhere-src/src/common/raft/integration/raft_initialization.cc line=53: call='cudaGetDeviceCount(&result)', Reason=cudaErrorInsufficientDriver:CUDA driver version is insufficient for CUDA runtime version

### Anything else?

See also the related discussion: https://github.com/milvus-io/milvus/discussions/36695

**Comment by ashkrisk** [comment]
I noticed that Milvus built from any of the 2.4.x branches has this issue, but Milvus built from the master branch does not. 

Turns out, the reason the master branch is not affected is because of the change in PR https://github.com/milvus-io/milvus/pull/34159. Here, the `gpu.initMemSize` and `gpu.maxMemSize` are assigned reasonable defaults.

It seemed odd that a simple configuration change could fix the issue, so I looked a little deeper, and it turns out there are actually two code branches that can be taken depending on if the these configurations are set or not:

https://github.com/milvus-io/milvus/blob/f3b6792a25a1ec8722f747ecbb44f9f384751c2f/internal/core/src/config/ConfigKnowhere.cpp#L97-L111

As seen in the above snippet, one of the two overrides of `knowhere::KnowhereConfig::SetRaftMemPool` will be called depending on if there are values provided for these configurations.

Looking into the Knowhere code for this function, we can see that in [one of the overrides,](https://github.com/zilliztech/knowhere/blob/be5e77c5127e7b08fc39ea4a9e1d269c8b6583d4/src/common/comp/knowhere_config.cc#L189) there is a sanity check present to make sure that there is a GPU available on the host machine
```cpp
    auto status = cudaGetDeviceCount(&count);
    // ...
    if (count < 1) {
        LOG_KNOWHERE_INFO_ << "GPU not available";
        return;
    }
```

But in the [other override](https://github.com/zilliztech/knowhere/blob/be5e77c5127e7b08fc39ea4a9e1d269c8b6583d4/src/common/comp/knowhere_config.cc#L212), there is no such check, which means the raft initialization will continue even when there is no GPU available, and inevitably fail.

**Comment by yanliang567** [comment]
/assign @Presburger 
please help to double check
/unassign

**Comment by Presburger** [comment]
> I noticed that Milvus built from any of the 2.4.x branches has this issue, but Milvus built from the master branch does not.
> 
> Turns out, the reason the master branch is not affected is because of the change in PR #34159. Here, the `gpu.initMemSize` and `gpu.maxMemSize` are assigned reasonable defaults.
> 
> It seemed odd that a simple configuration change could fix the issue, so I looked a little deeper, and it turns out there are actually two code branches that can be taken depending on if the these configurations are set or not:
> 
> https://github.com/milvus-io/milvus/blob/f3b6792a25a1ec8722f747ecbb44f9f384751c2f/internal/core/src/config/ConfigKnowhere.cpp#L97-L111
> 
> As seen in the above snippet, one of the two overrides of `knowhere::KnowhereConfig::SetRaftMemPool` will be called depending on if there are values provided for these configurations.
> 
> Looking into the Knowhere code for this function, we can see that in [one of the overrides,](https://github.com/zilliztech/knowhere/blob/be5e77c5127e7b08fc39ea4a9e1d269c8b6583d4/src/common/comp/knowhere_config.cc#L189) there is a sanity check present to make sure that there is a GPU available on the host machine
> 
> ```c++
>     auto status = cudaGetDeviceCount(&count);
>     // ...
>     if (count < 1) {
>         LOG_KNOWHERE_INFO_ << "GPU not available";
>         return;
>     }
> ```
> 
> But in the [other override](https://github.com/zilliztech/knowhere/blob/be5e77c5127e7b08fc39ea4a9e1d269c8b6583d4/src/common/comp/knowhere_config.cc#L212), there is no such check, which means the raft initialization will continue even when there is no GPU available, and inevitably fail.

The Milvus GPU memory pool cannot be set to `nullopt`, as this would result in significant memory wastage.

**Comment by Presburger** [comment]
For version 2.4.x, if you're sure this is the root cause, you can modify the `milvus.yaml` file to adjust the default memory pool size for the GPU. @ashkrisk

**Comment by ashkrisk** [comment]
> For version 2.4.x, if you're sure this is the root cause, you can modify the milvus.yaml file to adjust the default memory pool size for the GPU.

Yep, confirmed that the config change alone is enough to fix the issue on 2.4.x 👍

---

> The Milvus GPU memory pool cannot be set to nullopt, as this would result in significant memory wastage.

On the 2.4.x tags, isn't the memory pool size set to `nullopt` by default? In [this](https://github.com/zilliztech/knowhere/blob/db498ab36002b2886edb2b13f19fbc2ad89ec402/src/common/comp/knowhere_config.cc#L212) function:
```cpp
void
KnowhereConfig::SetRaftMemPool() {
    // Overload for default values
#ifdef KNOWHERE_WITH_RAFT
    auto config = raft_knowhere::raft_configuration{};
    raft_knowhere::initialize_raft(config);
#endif
}
```
`raft_configuration` is initialized without providing parameters, and the [defaults](https://github.com/zilliztech/knowhere/blob/db498ab36002b2886edb2b13f19fbc2ad89ec402/src/common/raft/integration/raft_initialization.hpp#L24) for this struct are all `nullopt`:
```cpp
struct raft_configuration {
    // ...
    std::optional<std::size_t> init_mem_pool_size_mb = std::nullopt;
    std::optional<std::size_t> max_mem_pool_size_mb = std::nullopt;
    std::optional<std::size_t> max_workspace_size_mb = std::nullopt;
};
```

Perhaps then the no-argument override for `SetRaftMemPool` can be changed to use some default byte values instead of `nullopt`? @Presburger

**Comment by Presburger** [comment]
@ashkrisk I will fix it today. The issue stems from the automated generation of the default YAML and the memory pool adjustments. These two commits did not match well, leading to the problem you encountered. Thank you very much for your feedback!
